### PR TITLE
buildRustPackage: support multiple `cargo` invocations

### PR DESCRIPTION
**Pull request by steveej**
###### Motivation for this change
This change allows calling `cargo build` multiple times with distinct
flags. Please see the doc changes for more details.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by symphorien** [comment]
what is the use case ?

**Comment by steveeJ** [comment]
> what is the use case ?

I just pushed a new commit to demonstrate using different features for different crates within the same workspace. Does that help understanding the use-case?

**Comment by symphorien** [comment]
I see. Just so we can test, can you name a crate/project where this is needed?

**Comment by steveeJ** [comment]
@symphorien of course. I'd like to integrate the changes made here with https://github.com/holochain/holochain/pull/438. The binaries there share a large portion of common dependencies so it is rather inefficient to use separate packages.

**Comment by lopsided98** [comment]
@steveeJ Have you seen [crate2nix](https://github.com/kolloch/crate2nix)? It might be able to solve your problem in a more elegant way.

**Comment by steveeJ** [comment]
@lopsided98 

> Have you seen [crate2nix](https://github.com/kolloch/crate2nix)? It might be able to solve your problem in a more elegant way.

I hadn't seen that, thanks. I would prefer if the mechanism lived in nixpkgs itself, so I'm suggesting that we do merge this PR. I'll also take a look at *crate2nix* nonetheless.

**Comment by steveeJ** [comment]
@lopsided98 @symphorien is there anything blocking this that I can help with?

**Comment by symphorien** [comment]
I must acknowledge that I don't really like this API. Possible suggestions:
* define a bash level function `runCargo` which runs cargo with all the default flags, and then it's possible to do
```nix
{
preBuild = ''
   runCargo --manifest-path foo bar
'';
}
```
* If the only possible use case is with different manifest paths, then make it possible to set `cargoBuildFlags` to a set with the path to the manifest as key and the flags as values.

What do you think?

**Comment by steveeJ** [comment]
@symphorien 

> I must acknowledge that I don't really like this API. Possible suggestions:
> (...)
> What do you think?

Well, I think we're unfortunately entering bike-shedding territory here. I prefer a Nix level solution or else I wouldn't have written it this way :wink: However, if you commit to approving the solution you outlined I'll implement it so we can get the functionality merged.

**Comment by symphorien** [comment]
Let's wait for another opinion then.

**Comment by otavio** [comment]
I think this is quite nice and indeed a nice addition to the system. I think using the nix system seems more ergonomic and natural to me. I support the @steveeJ proposal.

**Review by otavio — APPROVED** [review]

**Comment by steveeJ** [comment]
Thanks for the review @otavio. I rebased the PR and think we can merge this now. I haven't squashed the commits as I was previously told it makes reviewing harder. 

To the one who will merge this: please use the squash & merge button :slightly_smiling_face:

**Comment by SuperSandro2000** [comment]
> To the one who will merge this: please use the squash & merge button slightly_smiling_face

Just merge them together as the chances that someone forgets it are pretty high,

**Comment by steveeJ** [comment]
> Just squash them together as the chances that someone forgets it are pretty high.

Done :+1: 



> Also this needs to go through staging.

You mean I should target the `staging` branch rather than `master`?

**Comment by danieldk** [comment]
> You mean I should target the `staging` branch rather than `master`?

Yes, but that also requires quite some rewriting of this PR. `buildRustPackage` was completely rewritten to be hook-based. These changes are now in `staging-next` (they were merged to `staging`, because they require mass rebuilds). But until `staging-next` is merged into master, we shouldn't make any `buildRustPackage` changes in `master` to avoid merge conflicts.

File: pkgs/build-support/rust/default.nix
Comment (danieldk) [inline pkgs/build-support/rust/default.nix:38]: I agree with @symphorien . The `buildRustPackage` API is already quite complex as it is. Unless this use-case becomes very frequent in nixpkgs, I think it is better that projects that need to do multiple builds with different feature flags use their own `buildPhase`.

As @lopsided98 points out, such cases are very well handled by `buildRustCrate` plus `crate2nix`, since it compiles every crate as a separate output crate.

> I hadn't seen that, thanks. I would prefer if the mechanism lived in nixpkgs itself, so I'm suggesting that we do merge this PR. I'll also take a look at crate2nix nonetheless.

`crate2nix` is in nixpkgs and uses the `buildRustCrate` functionality in nixpkgs. If you do not want to add a `Cargo.nix` file in your repository and keep it up to date, `crate2nix` also supports IFD to generate the Nix from `Cargo.lock`.

**Review by danieldk — CHANGES_REQUESTED** [review]

**Comment by steveeJ** [comment]
i haven't been using this myself so i won't pursue this PR any longer

**Pull request closed without merging**
